lib/worker.js

```
@@ -0,0 +1,211 @@
+};
+
+Worker.prototype._doPostMessage = function(payload, transferList, type) {
+  this._checkAlive();
+
+  if (transferList == null)
+    transferList = EMPTY_ARRAY;
+  else if (!Array.isArray(transferList))
+    throw new Error('transferList must be an array');
+
+  this._workerContext.postMessage(JSONStringify(payload), transferList, type);
+};
+
+Worker.prototype.postMessage = function(payload, callback) {
+  this._doPostMessage(payload, null, WorkerContextBinding.USER);
+};
+
+Worker.prototype.terminate = function(callback) {
+  this._checkAlive();
+  var context = this._workerContext;
+  this._workerContext = null;
+  if (typeof callback === 'function') {
+    this.once('exit', function(exitCode) {
+      callback(null, exitCode);
+    });
+  }
+  context.terminate();
+};
+
+Worker.prototype.ref = function() {
+  this._checkAlive();
+  this._workerContext.ref();
+};
+
+Worker.prototype.unref = function() {
+  this._checkAlive();
+  this._workerContext.unref();
+};
+
+if (process.isWorkerInstance) {
+  const postMessage = function(payload, transferList, type) {
+    if (transferList == null)
+      transferList = EMPTY_ARRAY;
+    else if (!Array.isArray(transferList))
+      throw new Error('transferList must be an array');
+
+    WorkerContextBinding.workerWrapper._postMessage(JSONStringify(payload),
+                                                    transferList,
+                                                    type);
+  };
+  const workerFatalError = function(er) {
+    var defaultStack = null;
+    var defaultMessage = '[to string conversion failed]';
+    var defaultBuiltinType = 'Error';
+
+    var message = defaultMessage;
+    var builtinType = defaultBuiltinType;
+    var stack = defaultStack;
+    var additionalProperties = {};
+
+    if (er instanceof Error) {
+      try {
+        builtinType = er.name;
+        if (typeof builtinType !== 'string')
+          builtinType = defaultBuiltinType;
+      } catch (ignore) {}
```

Okay, but in that case you can move the 'is string' check out of the try/catch body.  Try/catch statements that swallow the exception are great places for bugs to hide, that's why I like to keep them as small as possible.